pkg/controllers/nodeclass/status/controller.go

```diff
@@ -0,0 +1,113 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package status
+
+import (
+	"context"
+
+	"go.uber.org/multierr"
+	"k8s.io/apimachinery/pkg/api/equality"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	corecontroller "sigs.k8s.io/karpenter/pkg/operator/controller"
+	"sigs.k8s.io/karpenter/pkg/utils/result"
+
+	"github.com/aws/karpenter-provider-aws/pkg/apis/v1beta1"
+	"github.com/aws/karpenter-provider-aws/pkg/providers/amifamily"
+	"github.com/aws/karpenter-provider-aws/pkg/providers/instanceprofile"
+	"github.com/aws/karpenter-provider-aws/pkg/providers/launchtemplate"
+	"github.com/aws/karpenter-provider-aws/pkg/providers/securitygroup"
+	"github.com/aws/karpenter-provider-aws/pkg/providers/subnet"
+)
+
+var _ corecontroller.TypedController[*v1beta1.EC2NodeClass] = (*Controller)(nil)
+
+type nodeClassStatusReconciler interface {
+	Reconcile(context.Context, *v1beta1.EC2NodeClass) (reconcile.Result, error)
+}
+
+type Controller struct {
+	kubeClient client.Client
+
+	ami             *AMI
+	instanceprofile *InstanceProfile
+	subnet          *Subnet
+	securitygroup   *SecurityGroup
+	launchtemplate  *LaunchTemplate
+}
+
+func NewController(kubeClient client.Client, subnetProvider subnet.Provider, securityGroupProvider securitygroup.Provider,
+	amiProvider amifamily.Provider, instanceProfileProvider instanceprofile.Provider, launchTemplateProvider launchtemplate.Provider) corecontroller.Controller {
+	return corecontroller.Typed[*v1beta1.EC2NodeClass](kubeClient, &Controller{
+		kubeClient: kubeClient,
+
+		ami:             &AMI{amiProvider: amiProvider},
+		subnet:          &Subnet{subnetProvider: subnetProvider},
+		securitygroup:   &SecurityGroup{securityGroupProvider: securityGroupProvider},
+		instanceprofile: &InstanceProfile{instanceProfileProvider: instanceProfileProvider},
+		launchtemplate:  &LaunchTemplate{launchTemplateProvider: launchTemplateProvider},
+	})
+}
+
+func (c *Controller) Reconcile(ctx context.Context, nodeClass *v1beta1.EC2NodeClass) (reconcile.Result, error) {
+	stored := nodeClass.DeepCopy()
```

Comment:
You should create this stored object after you patch, else you are going to be trying to patch the finalizer on two separate occassions